### ISSUE TITLE
test_tools_ftptools failed with ftplib.error_proto

### ISSUE THREAD

**Issue by dobedobedo** [remotesensinginfo/rsgislib#47]
Hi,  
When I tried to test the package with pytest, the test failed when calling the ftptools.upload_ftp_file function. Below is the error:
```
rsgislib/tools/ftptools.py:466: in upload_ftp_file
    ftp_conn.login(user=ftp_user, passwd=ftp_pass)
/usr/lib/python3.10/ftplib.py:414: in login
    resp = self.sendcmd('PASS ' + passwd)
/usr/lib/python3.10/ftplib.py:281: in sendcmd
    return self.getresp()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ftplib.FTP object at 0x7f4f4c56dab0>

    def getresp(self):
        resp = self.getmultiline()
        if self.debugging:
            print('*resp*', self.sanitize(resp))
        self.lastresp = resp[:3]
        c = resp[:1]
        if c in {'1', '2', '3'}:
            return resp
        if c == '4':
            raise error_temp(resp)
        if c == '5':
            raise error_perm(resp)
>       raise error_proto(resp)
E       ftplib.error_proto: /usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'

/usr/lib/python3.10/ftplib.py:255: error_proto
```
Below is the full log of pytest:  
<details>
<summary>pytest log</summary>

```
==================================================================================================================================== test session starts ====================================================================================================================================
platform linux -- Python 3.10.4, pytest-7.1.1, pluggy-1.0.0
rootdir: /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test
plugins: cov-2.12.1
collected 884 items                                                                                                                                                                                                                                                                         

tests/test_changedetect_pxloutlierchng.py s...                                                                                                                                                                                                                                        [  0%]
tests/test_classification.py ................                                                                                                                                                                                                                                         [  2%]
tests/test_classification_classaccuracymetrics.py ........                                                                                                                                                                                                                            [  3%]
tests/test_classification_classimblearn.py sssssssssssssss                                                                                                                                                                                                                            [  4%]
tests/test_classification_classlightgbm.py ssssss                                                                                                                                                                                                                                     [  5%]
tests/test_classification_classratutils.py ..                                                                                                                                                                                                                                         [  5%]
tests/test_classification_classsklearn.py ....                                                                                                                                                                                                                                        [  6%]
tests/test_classification_classxgboost.py sssssssss                                                                                                                                                                                                                                   [  7%]
tests/test_classification_clustersklearn.py ...                                                                                                                                                                                                                                       [  7%]
tests/test_elevation.py .............                                                                                                                                                                                                                                                 [  9%]
tests/test_image_morphology.py .............                                                                                                                                                                                                                                          [ 10%]
tests/test_imagecalc.py ..........................................................                                                                                                                                                                                                    [ 17%]
tests/test_imagecalc_calcindices.py .............                                                                                                                                                                                                                                     [ 18%]
tests/test_imagecalc_leastcostpath.py s                                                                                                                                                                                                                                               [ 18%]
tests/test_imagecalc_specunmixing.py ............ssssssss.....                                                                                                                                                                                                                        [ 21%]
tests/test_imagefilter.py .............................................................                                                                                                                                                                                               [ 28%]
tests/test_imageregistration.py ......                                                                                                                                                                                                                                                [ 29%]
tests/test_imageutils.py .....................................................................................................................................                                                                                                                        [ 44%]
tests/test_imageutils_imagelut.py ........                                                                                                                                                                                                                                            [ 45%]
tests/test_imageutils_stdimgblockiter.py ..                                                                                                                                                                                                                                           [ 45%]
tests/test_rastergis.py ........ss.................                                                                                                                                                                                                                                   [ 48%]
tests/test_regression.py .                                                                                                                                                                                                                                                            [ 48%]
tests/test_regression_regresssklearn.py .....                                                                                                                                                                                                                                         [ 48%]
tests/test_rsgislib_cli.py ........ssss.....................................s....                                                                                                                                                                                                     [ 55%]
tests/test_segentatation.py ...                                                                                                                                                                                                                                                       [ 55%]
tests/test_segentatation_shepherdseg.py ....                                                                                                                                                                                                                                          [ 55%]
tests/test_segentatation_skimgseg.py ssss                                                                                                                                                                                                                                             [ 56%]
tests/test_segentatation_tiledclump.py ....                                                                                                                                                                                                                                           [ 56%]
tests/test_timeseries_modelfitting.py ..                                                                                                                                                                                                                                              [ 57%]
tests/test_tools_checkdatasets.py .....................                                                                                                                                                                                                                               [ 59%]
tests/test_tools_filetools.py ..................................ss........                                                                                                                                                                                                            [ 64%]
tests/test_tools_ftptools.py FssFFss                                                                                                                                                                                                                                                  [ 65%]
tests/test_tools_geometrytools.py .......                                                                                                                                                                                                                                             [ 65%]
tests/test_tools_plotting.py ..............                                                                                                                                                                                                                                           [ 67%]
tests/test_tools_projection.py .........                                                                                                                                                                                                                                              [ 68%]
tests/test_tools_utils.py ..............................................................                                                                                                                                                                                              [ 75%]
tests/test_tools_visualisation.py s......ss.sss..ssssssss.                                                                                                                                                                                                                            [ 78%]
tests/test_vectorattrs.py .........................                                                                                                                                                                                                                                   [ 81%]
tests/test_vectorgeoms.py ..................s.......................                                                                                                                                                                                                                  [ 85%]
tests/test_vectorutils.py .....................................................................                                                                                                                                                                                       [ 93%]
tests/test_vectorutils_createrasters.py ......                                                                                                                                                                                                                                        [ 94%]
tests/test_vectorutils_createvectors.py .............                                                                                                                                                                                                                                 [ 95%]
tests/test_zonalstats.py .....................................                                                                                                                                                                                                                        [100%]

========================================================================================================================================= FAILURES ==========================================================================================================================================
___________________________________________________________________________________________________________________________________ test_upload_ftp_file ____________________________________________________________________________________________________________________________________

    def test_upload_ftp_file():
        import rsgislib.tools.ftptools
    
        test_file = os.path.join(TOOLS_UTILS_DATA_DIR, "basic_str.txt")
>       rsgislib.tools.ftptools.upload_ftp_file(
            test_file,
            ftp_url=ftp_url_ref,
            remote_path="/rsgislib_test/test",
            ftp_timeout=300,
            ftp_user=ftp_user_ref,
            ftp_pass=ftp_pass_ref,
            print_info=True,
        )

tests/test_tools_ftptools.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
rsgislib/tools/ftptools.py:466: in upload_ftp_file
    ftp_conn.login(user=ftp_user, passwd=ftp_pass)
/usr/lib/python3.10/ftplib.py:414: in login
    resp = self.sendcmd('PASS ' + passwd)
/usr/lib/python3.10/ftplib.py:281: in sendcmd
    return self.getresp()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ftplib.FTP object at 0x7f4f5921c8b0>

    def getresp(self):
        resp = self.getmultiline()
        if self.debugging:
            print('*resp*', self.sanitize(resp))
        self.lastresp = resp[:3]
        c = resp[:1]
        if c in {'1', '2', '3'}:
            return resp
        if c == '4':
            raise error_temp(resp)
        if c == '5':
            raise error_perm(resp)
>       raise error_proto(resp)
E       ftplib.error_proto: /usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'

/usr/lib/python3.10/ftplib.py:255: error_proto
----------------------------------------------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------------------------------------------------
Opening FTP Connection to ftp.dlptest.com
__________________________________________________________________________________________________________________________________ test_download_ftp_file ___________________________________________________________________________________________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-o0331dobe/pytest-2/test_download_ftp_file0')

    def test_download_ftp_file(tmp_path):
        import rsgislib.tools.ftptools
    
        test_file = os.path.join(TOOLS_UTILS_DATA_DIR, "basic_str.txt")
>       rsgislib.tools.ftptools.upload_ftp_file(
            test_file,
            ftp_url=ftp_url_ref,
            remote_path="/rsgislib_test/test",
            ftp_timeout=300,
            ftp_user=ftp_user_ref,
            ftp_pass=ftp_pass_ref,
            print_info=True,
        )

tests/test_tools_ftptools.py:101: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
rsgislib/tools/ftptools.py:466: in upload_ftp_file
    ftp_conn.login(user=ftp_user, passwd=ftp_pass)
/usr/lib/python3.10/ftplib.py:414: in login
    resp = self.sendcmd('PASS ' + passwd)
/usr/lib/python3.10/ftplib.py:281: in sendcmd
    return self.getresp()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ftplib.FTP object at 0x7f4f4c56dab0>

    def getresp(self):
        resp = self.getmultiline()
        if self.debugging:
            print('*resp*', self.sanitize(resp))
        self.lastresp = resp[:3]
        c = resp[:1]
        if c in {'1', '2', '3'}:
            return resp
        if c == '4':
            raise error_temp(resp)
        if c == '5':
            raise error_perm(resp)
>       raise error_proto(resp)
E       ftplib.error_proto: /usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'

/usr/lib/python3.10/ftplib.py:255: error_proto
----------------------------------------------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------------------------------------------------
Opening FTP Connection to ftp.dlptest.com
________________________________________________________________________________________________________________________________ test_download_curl_ftp_file ________________________________________________________________________________________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-o0331dobe/pytest-2/test_download_curl_ftp_file0')

    @pytest.mark.skipif(PYCURL_DB_NOT_AVAIL, reason="pycurl dependency not available")
    def test_download_curl_ftp_file(tmp_path):
        import rsgislib.tools.ftptools
    
        test_file = os.path.join(TOOLS_UTILS_DATA_DIR, "basic_str.txt")
>       rsgislib.tools.ftptools.upload_ftp_file(
            test_file,
            ftp_url=ftp_url_ref,
            remote_path="/rsgislib_test/test",
            ftp_timeout=300,
            ftp_user=ftp_user_ref,
            ftp_pass=ftp_pass_ref,
            print_info=True,
        )

tests/test_tools_ftptools.py:129: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
rsgislib/tools/ftptools.py:466: in upload_ftp_file
    ftp_conn.login(user=ftp_user, passwd=ftp_pass)
/usr/lib/python3.10/ftplib.py:414: in login
    resp = self.sendcmd('PASS ' + passwd)
/usr/lib/python3.10/ftplib.py:281: in sendcmd
    return self.getresp()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ftplib.FTP object at 0x7f4f4c5068c0>

    def getresp(self):
        resp = self.getmultiline()
        if self.debugging:
            print('*resp*', self.sanitize(resp))
        self.lastresp = resp[:3]
        c = resp[:1]
        if c in {'1', '2', '3'}:
            return resp
        if c == '4':
            raise error_temp(resp)
        if c == '5':
            raise error_perm(resp)
>       raise error_proto(resp)
E       ftplib.error_proto: /usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'

/usr/lib/python3.10/ftplib.py:255: error_proto
----------------------------------------------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------------------------------------------------
Opening FTP Connection to ftp.dlptest.com
===================================================================================================================================== warnings summary ======================================================================================================================================
../../../../../../../../usr/lib/python3.10/site-packages/geopandas/_compat.py:2
  /usr/lib/python3.10/site-packages/geopandas/_compat.py:2: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.version import LooseVersion

tests/test_changedetect_pxloutlierchng.py::test_find_class_kurt_skew_outliers
  /usr/lib/python3.10/site-packages/rios/imageio.py:63: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    (numpy.bool,gdalconst.GDT_Byte),

tests/test_changedetect_pxloutlierchng.py::test_find_class_kurt_skew_outliers
  /usr/lib/python3.10/site-packages/rios/imageio.py:69: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    (numpy.float,gdalconst.GDT_Float64)

tests/test_classification.py::test_get_class_training_data
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorutils/__init__.py:252: DeprecationWarning: invalid escape sequence '\*'
    """

tests/test_classification.py::test_get_class_training_data
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorutils/__init__.py:334: DeprecationWarning: invalid escape sequence '\*'
    """

tests/test_classification.py::test_get_class_training_data
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorutils/__init__.py:428: DeprecationWarning: invalid escape sequence '\*'
    """

tests/test_classification.py: 10 warnings
tests/test_classification_classaccuracymetrics.py: 4 warnings
tests/test_vectorattrs.py: 9 warnings
tests/test_vectorgeoms.py: 9 warnings
tests/test_vectorutils.py: 59 warnings
tests/test_vectorutils_createrasters.py: 1 warning
tests/test_vectorutils_createvectors.py: 1 warning
  /usr/lib/python3.10/site-packages/geopandas/io/file.py:362: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
    pd.Int64Index,

tests/test_classification_classaccuracymetrics.py: 10 warnings
tests/test_vectorutils.py: 3 warnings
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorutils/__init__.py:2765: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
    data_gdf = data_gdf.append(tmp_data_gdf)

tests/test_regression_regresssklearn.py::test_perform_search_param_opt
  /usr/lib/python3.10/site-packages/sklearn/model_selection/_search.py:292: UserWarning: The total space of parameters 1 is smaller than n_iter=25. Running 1 iterations. For exhaustive searches, use GridSearchCV.
    warnings.warn(

tests/test_regression_regresssklearn.py::test_perform_search_param_opt
tests/test_regression_regresssklearn.py::test_perform_search_param_opt
  /usr/lib/python3.10/site-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
    estimator.fit(X_train, y_train, **fit_params)

tests/test_regression_regresssklearn.py::test_perform_search_param_opt
  /usr/lib/python3.10/site-packages/sklearn/model_selection/_search.py:926: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
    self.best_estimator_.fit(X, y, **fit_params)

tests/test_regression_regresssklearn.py::test_perform_kfold_fit
tests/test_regression_regresssklearn.py::test_perform_kfold_fit
tests/test_regression_regresssklearn.py::test_perform_kfold_fit
tests/test_regression_regresssklearn.py::test_perform_kfold_fit
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/regression/regresssklearn.py:273: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
    skl_regrs_obj.fit(x_train, y_train)

tests/test_regression_regresssklearn.py::test_apply_regress_sklearn_mdl
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/tests/test_regression_regresssklearn.py:112: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
    et_obj.fit(x, y)

tests/test_vectorgeoms.py::test_explode_vec_lyr
tests/test_vectorgeoms.py::test_explode_vec_files
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorgeoms/__init__.py:1634: FutureWarning: Currently, index_parts defaults to True, but in the future, it will default to False to be consistent with Pandas. Use `index_parts=True` to keep the current behavior and True/False to silence the warning.
    data_explode_gdf = data_gdf.explode()

tests/test_vectorutils.py::test_perform_spatial_join_empty
tests/test_vectorutils.py::test_perform_spatial_join
  /usr/lib/python3.10/site-packages/_pytest/python.py:192: FutureWarning: The `op` parameter is deprecated and will be removed in a future release. Please use the `predicate` parameter instead.
    result = testfunction(**testargs)

tests/test_vectorutils.py::test_merge_vector_files
tests/test_vectorutils.py::test_merge_vector_files
tests/test_vectorutils.py::test_merge_vector_files
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorutils/__init__.py:2711: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
    data_gdf = data_gdf.append(tmp_data_gdf)

tests/test_vectorutils.py::test_merge_utm_vecs_wgs84
tests/test_vectorutils.py::test_merge_utm_vecs_wgs84
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorutils/__init__.py:3019: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
    out_gdf = out_gdf.append(data_gdf)

tests/test_vectorutils.py::test_merge_utm_vecs_wgs84
  /home/o0331dobe/Documents/AUR_packages/python-rsgislib/src/rsgislib-5.0.7/build-test/rsgislib/vectorutils/__init__.py:2852: ShapelyDeprecationWarning: Iteration over multi-part geometries is deprecated and will be removed in Shapely 2.0. Use the `geoms` property to access the constituent parts of a multi-part geometry.
    for poly in row["geometry"]:

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.10.4-final-0 -----------
Name                                              Stmts   Miss  Cover
---------------------------------------------------------------------
rsgislib/__init__.py                                283     86    70%
rsgislib/changedetect/__init__.py                     1      0   100%
rsgislib/changedetect/pxloutlierchng.py             170     95    44%
rsgislib/classification/__init__.py                 404    177    56%
rsgislib/classification/classaccuracymetrics.py     889     46    95%
rsgislib/classification/classimblearn.py            109    109     0%
rsgislib/classification/classkeraschips.py          187    187     0%
rsgislib/classification/classkeraspxl.py            155    155     0%
rsgislib/classification/classlightgbm.py            467    467     0%
rsgislib/classification/classratutils.py             62      0   100%
rsgislib/classification/classsklearn.py             263     14    95%
rsgislib/classification/classxgboost.py             691    691     0%
rsgislib/classification/clustersklearn.py           112     10    91%
rsgislib/dataaccess/__init__.py                       1      1     0%
rsgislib/dataaccess/usgs_m2m.py                     242    242     0%
rsgislib/elevation/__init__.py                       42      9    79%
rsgislib/imagecalc/__init__.py                      692    119    83%
rsgislib/imagecalc/calcindices.py                   133     13    90%
rsgislib/imagecalc/leastcostpath.py                  51     51     0%
rsgislib/imagecalc/specunmixing/__init__.py         606    302    50%
rsgislib/imagecalibration/__init__.py               115    115     0%
rsgislib/imagecalibration/sensorlvl2data.py          64     64     0%
rsgislib/imagecalibration/solarangles.py             56     56     0%
rsgislib/imagefilter/__init__.py                    188      0   100%
rsgislib/imagefilter/tiledfilter.py                 444    155    65%
rsgislib/imagemorphology/__init__.py                  2      0   100%
rsgislib/imageregistration/__init__.py               53     15    72%
rsgislib/imageutils/__init__.py                    1645    301    82%
rsgislib/imageutils/imagecomp.py                    513    513     0%
rsgislib/imageutils/imagelut.py                      99      9    91%
rsgislib/imageutils/stdimgblockiter.py              171     22    87%
rsgislib/imageutils/tilingutils.py                  109    109     0%
rsgislib/rastergis/__init__.py                      458    288    37%
rsgislib/regression/__init__.py                      41      3    93%
rsgislib/regression/regresssklearn.py               184     16    91%
rsgislib/segmentation/__init__.py                     2      0   100%
rsgislib/segmentation/shepherdseg.py                189     27    86%
rsgislib/segmentation/skimgseg.py                   251    251     0%
rsgislib/segmentation/tiledclump.py                 167     16    90%
rsgislib/segmentation/tiledsegsingle.py             252    252     0%
rsgislib/timeseries/__init__.py                       1      0   100%
rsgislib/timeseries/modelfitting.py                 195     24    88%
rsgislib/timeseries/robustfitoutliners.py           137    137     0%
rsgislib/timeseries/tmask.py                        102    102     0%
rsgislib/tools/__init__.py                            1      0   100%
rsgislib/tools/checkdatasets.py                     410    158    61%
rsgislib/tools/filetools.py                         556    257    54%
rsgislib/tools/ftptools.py                          209    189    10%
rsgislib/tools/geometrytools.py                     238     68    71%
rsgislib/tools/googlecloud.py                        26     26     0%
rsgislib/tools/httptools.py                          83     83     0%
rsgislib/tools/plotting.py                          425    232    45%
rsgislib/tools/projection.py                         57      4    93%
rsgislib/tools/sensors.py                            16     16     0%
rsgislib/tools/stats.py                             389    274    30%
rsgislib/tools/sysprofile.py                         45     45     0%
rsgislib/tools/testimages.py                         30     30     0%
rsgislib/tools/tilecacheutils.py                    187    134    28%
rsgislib/tools/utils.py                             359    170    53%
rsgislib/tools/utm.py                               209    120    43%
rsgislib/tools/visualisation.py                     791    509    36%
rsgislib/vectorattrs/__init__.py                    379     58    85%
rsgislib/vectorgeoms/__init__.py                   1738    371    79%
rsgislib/vectorutils/__init__.py                   1407    253    82%
rsgislib/vectorutils/createrasters.py                77     23    70%
rsgislib/vectorutils/createvectors.py               496    159    68%
rsgislib/zonalstats/__init__.py                    1320    287    78%
---------------------------------------------------------------------
TOTAL                                             20446   8715    57%

================================================================================================================================== short test summary info ==================================================================================================================================
FAILED tests/test_tools_ftptools.py::test_upload_ftp_file - ftplib.error_proto: /usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'
FAILED tests/test_tools_ftptools.py::test_download_ftp_file - ftplib.error_proto: /usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'
FAILED tests/test_tools_ftptools.py::test_download_curl_ftp_file - ftplib.error_proto: /usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'
============================================================================================================ 3 failed, 809 passed, 72 skipped, 131 warnings in 327.20s (0:05:27) ============================================================================================================
```
</details>

**Comment by petebunting** [comment]
Thanks for this. How did you install RSGISLib? Did you install via conda / conda-forge? 

Looks like the error is coming from the python installation dependencies. What happens if you just import the ftplib from the python prompt?

**Comment by dobedobedo** [comment]
Hi,
I built the package from source with the make arguments:  
```
cmake \
    -D Python_ROOT_DIR:FILEPATH=/usr \
    -D Python_FIND_STRATEGY=LOCATION \
    -D CMAKE_FIND_FRAMEWORK=NEVER \
    -D CMAKE_PREFIX_PATH=/usr \
    -D CMAKE_INSTALL_PREFIX=/usr \
    -D BOOST_INCLUDE_DIR=/usr/include \
    -D BOOST_LIB_PATH=/usr/lib \
    -D GDAL_INCLUDE_DIR=/usr/include \
    -D GDAL_LIB_PATH=/usr/lib \
    -D HDF5_INCLUDE_DIR=/usr/include \
    -D HDF5_LIB_PATH=/usr/lib \
    -D GSL_INCLUDE_DIR=/usr/include \
    -D GSL_LIB_PATH=/usr/lib \
    -D MUPARSER_INCLUDE_DIR=/usr/include \
    -D MUPARSER_LIB_PATH=/usr/lib \
    -D KEA_INCLUDE_DIR=/usr/include \
    -D KEA_LIB_PATH=/usr/lib \
    -D CMAKE_BUILD_TYPE=Release \
    -D CMAKE_SKIP_RPATH=ON
  make
```
However, I found that it returns the same message when I used `ftp` command in the terminal with the pytest FTP profiles:  
```
Connected to ftp.dlptest.com.
220 Welcome to the DLP Test FTP Server
Name (ftp.dlptest.com:o0331dobe): dlpuser
331 Please specify the password.
Password: 
/usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'
ftp: Login failed.
ftp> bye
530 Please login with USER and PASS.
```
So the problem might not be the Python issue but something else.

**Comment by dobedobedo** [comment]
Hi,
I thought it was because I didn't install vsftpd, but after I installed and enable the service, it still shows the same result:
```
ftp ftp.dlptest.com                                                                                                                                                                                                                                                           ✔ 
Connected to ftp.dlptest.com.
220 Welcome to the DLP Test FTP Server
Name (ftp.dlptest.com:o0331dobe): dlpuser
331 Please specify the password.
Password: 
/usr/sbin/vsftpd: Relink `/lib64/libnss_myhostname.so.2' with `/lib64/librt.so.1' for IFUNC symbol `clock_gettime'
ftp: Login failed.
Remote system type is Login.
ftp> bye
215 UNIX Type: L8
```

**Comment by petebunting** [comment]
Hi,

Well the error is coming from the vsftpd build on your system and not rsgislib so I guess I would have a couple of questions/comments:

1. do you need the ftp tools which are in rsgislib? They are just a few python functions for helping to download data from ftp servers so probably not. In which case you can just ignore this error and get on as this error won’t effect you.
2. have you tried the conda-forge build? That is the easiest way to install rsgislib and might be easier?
3. Have you googled the error message? From the quick search I just did it sounds like it might even be a error which occurs when there is a firewall issue (I.e., the ftp server can’t be reached). If you really want to check the issue out you might want to try just accessing the test service I’m using here though the terminal ftp client and see what happens. If there is another ftp service which you can access you could change the test to you that server.

**Comment by dobedobedo** [comment]
Thanks for the suggestions. I probably don't need the FTP functions so I can probably live with it. I tried the conda build on another Ubuntu machine, but the pytest on ftptool still ended with the same result, so that's why I speculate that might be the FTP server issue, though I couldn't find other public FTP services for testing. I tried disabling my system firewall too but ended with the same result.
